sql/schema.go

```diff
@@ -0,0 +1,202 @@
+package sql
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	godiffpatch "github.com/sourcegraph/go-diff-patch"
+	"go.uber.org/zap"
+)
+
+const (
+	SchemaPath        = "schema/schema.sql"
+	UpdatedSchemaPath = "schema/schema.sql.updated"
+)
+
+// LoadSchema loads the schema embedded in the executable.
+func LoadSchema(fsys fs.FS, migrations []Migration) (*Schema, error) {
+	text, err := fs.ReadFile(fsys, SchemaPath)
+	if err != nil {
+		return nil, fmt.Errorf("error reading schema file %s: %w", SchemaPath, err)
+	}
+	return &Schema{Script: string(text), Migrations: migrations}, nil
+}
+
+// LoadDBSchemaScript retrieves the database schema as text.
+func LoadDBSchemaScript(db Executor, ignoreRx string) (string, error) {
+	var (
+		err   error
+		ignRx *regexp.Regexp
+		sb    strings.Builder
+	)
+	if ignoreRx != "" {
+		ignRx, err = regexp.Compile(ignoreRx)
+		if err != nil {
+			return "", fmt.Errorf("error compiling table ignore regexp %q: %w",
+				ignoreRx, err)
+		}
+	}
+	version, err := version(db)
+	if err != nil {
+		return "", err
+	}
+	fmt.Fprintf(&sb, "PRAGMA user_version = %d;\n", version)
+	if _, err = db.Exec(
+		// Type is either 'index' or 'table', we want tables
+		// to go first. Also, we ignore _litestream tables
+		`select tbl_name, sql || ';' from sqlite_master
+                 where sql is not null
+                 order by tbl_name, type desc, name`,
+		nil, func(st *Statement) bool {
+			if ignRx == nil || !ignRx.MatchString(st.ColumnText(0)) {
+				fmt.Fprintln(&sb, st.ColumnText(1))
+			}
+			return true
+		}); err != nil {
+		return "", fmt.Errorf("error retrieving DB schema: %w", err)
+	}
+	return sb.String(), nil
+}
+
+// Schema represents database schema.
+type Schema struct {
+	Script        string
+	Migrations    MigrationList
+	skipMigration map[int]struct{}
+}
+
+// Diff diffs the database schema against the actual schema.
+// If there's no differences, it returns an empty string.
+func (s *Schema) Diff(actualScript string) string {
+	if s.Script == actualScript {
+		return ""
+	}
+	diff := godiffpatch.GeneratePatch(SchemaPath, s.Script, actualScript)
+	if diff == "" {
+		return "<diff>"
+	}
+	return diff
+}
+
+// WriteToFile writes the schema to the corresponding updated schema file.
+func (s *Schema) WriteToFile(basedir string) error {
+	path := filepath.Join(basedir, UpdatedSchemaPath)
+	if err := os.WriteFile(path, []byte(s.Script), 0o777); err != nil {
+		return fmt.Errorf("error writing schema file %s: %w", path, err)
+	}
+	return nil
```

Comment:
This function seems to only exist for tests and there it is also not used besides writing it to a file and logging that it has been written? Example from the test:

```go
			loadedScript, err := sql.LoadDBSchemaScript(db, "")
			require.NoError(t, err)
			expSchema, err := Schema()
			require.NoError(t, err)
			diff := expSchema.Diff(loadedScript)
			if diff != "" {
				s := &sql.Schema{Script: loadedScript}
				require.NoError(t, s.WriteToFile("."))
				t.Logf("updated schema written to %s", sql.UpdatedSchemaPath)
			}
			require.Empty(t, diff, "schema diff")
```

That could just be:

```go
			loadedScript, err := sql.LoadDBSchemaScript(db, "")
			require.NoError(t, err)
			expSchema, err := Schema()
			require.NoError(t, err)
			require.Equal(t, expSchema.Script, loadedScript)
```

This already prints a diff if they are not the same and doesn't require `Schema.Diff` and `Schema.WriteToFile`

Reply:
`WriteFile` is needed so that the updated file can be used to replace schema file easily.
The expected workflow is this: you never edit schema file. Instead, you add migrations.
After you do that the the test fails and it writes the updated file, which you can rename to `schema.sql` if changes there are what you expect after migration. After you rename the updated schema file to `schema.sql` the test should pass (unless you have flaky migration that does different schema changes each time you run it)

Reply:
Tried doing just `require.Equal`, but this prints out the whole script in the unreadable escaped form (`\n` instead of newlines) before printing the diff, which makes test output much more cluttered

Reply:
I manually edited the `schema.sql` and executed the tests here is the output:

```
--- FAIL: TestSchema (0.04s)
    --- FAIL: TestSchema/idempotent_migration (0.02s)
        /workspaces/spacemesh/go-spacemesh/sql/statesql/logger.go:146: 2024-06-05T20:34:47.296Z	INFO	running migrations	{"uri": "file:/tmp/TestSchemaidempotent_migration785241580/001/test.db", "current version": 0, "target version": 18}
        /workspaces/spacemesh/go-spacemesh/sql/statesql/logger.go:146: 2024-06-05T20:34:47.310Z	WARN	database schema drift detected	{"uri": "file:/tmp/TestSchemaidempotent_migration785241580/001/test.db", "diff": "  (\n  \t\"\"\"\n  \t... // 9 identical lines\n  \t    PRIMARY KEY (address, layer_updated DESC)\n  \t);\n+ \tCREATE INDEX accounts_by_layer_updated ON accounts (layer_updated);\n  \tCREATE TABLE activesets\n  \t(\n  \t... // 138 identical lines\n  \t\"\"\"\n  )\n"}
        /workspaces/spacemesh/go-spacemesh/sql/statesql/test.go:52: 
            	Error Trace:	/workspaces/spacemesh/go-spacemesh/sql/test/test.go:52
            	Error:      	Not equal: 
            	            	expected: 1
            	            	actual  : 2
            	Test:       	TestSchema/idempotent_migration
            	Messages:   	expected 1 log messages
    --- FAIL: TestSchema/schema (0.02s)
        --- FAIL: TestSchema/schema/no_migrations (0.01s)
            /workspaces/spacemesh/go-spacemesh/sql/statesql/test.go:102: 
                	Error Trace:	/workspaces/spacemesh/go-spacemesh/sql/test/test.go:102
                	Error:      	Not equal: 
                	            	expected: ***abbreviated***
                	            	actual  : ***abbreviated***
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -11,2 +11,3 @@
                	            	 );
                	            	+CREATE INDEX accounts_by_layer_updated ON accounts (layer_updated);
                	            	 CREATE TABLE activesets
                	Test:       	TestSchema/schema/no_migrations
        --- FAIL: TestSchema/schema/force_migrations (0.01s)
            /workspaces/spacemesh/go-spacemesh/sql/statesql/test.go:102: 
                	Error Trace:	/workspaces/spacemesh/go-spacemesh/sql/test/test.go:102
                	Error:      	Not equal: 
                	            	expected: ***abbreviated***
                	            	actual  : ***abbreviated***
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -11,2 +11,3 @@
                	            	 );
                	            	+CREATE INDEX accounts_by_layer_updated ON accounts (layer_updated);
                	            	 CREATE TABLE activesets
                	Test:       	TestSchema/schema/force_migrations
FAIL
```

Yes it prints the full values of the strings (here abbreviated) but also just the diff.

Reply:
> The expected workflow is this: you never edit schema file. Instead, you add migrations.
After you do that the the test fails and it writes the updated file, which you can rename to schema.sql if changes there are what you expect after migration.

I'm not sure I like this workflow. We had such tests before where the test (or in this case here the shipped schema) needs to be updated by hand after adding a new migration. I'm not sure however what the alternative should be. In another project we had a dedicated tool that would generate up and down migrations for different DBs based on a DSL we wrote the migrations in and it would also generate the expected schema of the fully migrated DB when executed. Maybe the test is an OK substitute for this at the moment 🤔

Reply:
Switched to a workflow that uses `make generate`